src/vanilla/store.ts

```diff
@@ -225,6 +227,16 @@ export const createStore = () => {
     abortPromise?: () => void
   ): AtomState<Value> => {
     if (valueOrPromise instanceof Promise) {
+      const prevAtomState = getAtomState(atom)
+      if (prevAtomState?.o === valueOrPromise) {
+        return setAtomValue(
+          atom,
+          prevAtomState.v as Value,
+          nextDependencies,
```

Comment:
This looks bad (it clears up dependencies). I'll think about the solution.

Reply:
The dependencies could have changed even if the promise is the same

Reply:
If dependencies are changed, we need to set it after the promise resolution, no?

Reply:
The async deps most likely can not change (they can with some creativity, but let's consider this out of scope for now). But the sync deps can change

Reply:
async deps include sync deps, so they can change. I'm afraid this isn't solvable easily.

Reply:
Added some commits. Hope it works.